src/Type/Php/CallUserFuncDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,32 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\Type;
+use function array_slice;
+
+class CallUserFuncDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
+	{
+		$args = $functionCall->getArgs();
+
+		$callbackType = $scope->getType($args[0]->value);
+		if ($callbackType->isCallable()->yes()) {
+			return $scope->getType(new FuncCall($args[0]->value, array_slice($args, 1)));
```

Comment:
I'd say this is a bit naive about named arguments. See https://wiki.php.net/rfc/named_params - section "call_user_func() and friends".

 Most likely we'll have to modify ArgumentsNormalizer as well.